sklearn/_fast_gradient_boosting/splitting.pyx

```diff
@@ -0,0 +1,734 @@
+        (min_gain_to_split, etc.) are discarded here. If no split can
+        satisfy the constraints, a SplitInfo with a gain of -1 is returned.
+        If for a given node the best SplitInfo has a gain of -1, it is
+        finalized into a leaf in the grower.
+        """
+        cdef:
+            unsigned int bin_idx
+            unsigned int n_samples_left
+            unsigned int n_samples_right
+            unsigned int n_samples_ = n_samples
+            Y_DTYPE_C sum_hessian_left
+            Y_DTYPE_C sum_hessian_right
+            Y_DTYPE_C sum_gradient_left
+            Y_DTYPE_C sum_gradient_right
+            Y_DTYPE_C gain
+            split_info_struct best_split
+
+        best_split.gain = -1.
+        sum_gradient_left, sum_hessian_left = 0., 0.
+        n_samples_left = 0
+
+        for bin_idx in range(self.n_bins_per_feature[feature_idx]):
+            n_samples_left += histograms[feature_idx, bin_idx].count
+            n_samples_right = n_samples_ - n_samples_left
+
+            if self.hessians_are_constant:
+                sum_hessian_left += histograms[feature_idx, bin_idx].count
+            else:
+                sum_hessian_left += \
+                    histograms[feature_idx, bin_idx].sum_hessians
+            sum_hessian_right = sum_hessians - sum_hessian_left
+
+            sum_gradient_left += histograms[feature_idx, bin_idx].sum_gradients
+            sum_gradient_right = sum_gradients - sum_gradient_left
+
+            if n_samples_left < self.min_samples_leaf:
+                continue
+            if n_samples_right < self.min_samples_leaf:
+                # won't get any better
+                break
+
+            if sum_hessian_left < self.min_hessian_to_split:
+                continue
+            if sum_hessian_right < self.min_hessian_to_split:
+                # won't get any better (hessians are > 0 since loss is convex)
+                break
+
+            gain = _split_gain(sum_gradient_left, sum_hessian_left,
+                               sum_gradient_right, sum_hessian_right,
+                               sum_gradients, sum_hessians,
+                               self.l2_regularization)
+
+            if gain - best_split.gain > EPS and gain > self.min_gain_to_split:
+                best_split.gain = gain
+                best_split.feature_idx = feature_idx
+                best_split.bin_idx = bin_idx
+                best_split.sum_gradient_left = sum_gradient_left
+                best_split.sum_gradient_right = sum_gradient_right
+                best_split.sum_hessian_left = sum_hessian_left
+                best_split.sum_hessian_right = sum_hessian_right
+                best_split.n_samples_left = n_samples_left
+                best_split.n_samples_right = n_samples_right
+
+        return best_split
+
+    # Only used for tests (python code cannot use cdef types)
+    # Not sure if this is a good practice...
+    def find_best_split_wrapper(
```

Comment:
Making this "private" by prefixing `find_best_split_wrapper` with `_` would discourage people from using this.